### DISCUSSION TITLE
Example with GRIB2 data

### DISCUSSION THREAD

**Discussion by blaylockbk** [opengeos/leafmap#229] [Q&A] [answered]
There is a lot of weather data in the GRIB2 format. It's easy to read GRIB2 into an xarray DataFrame with cfgrib. Is there an easy way to add data from xarray to leafmap? Sorry if I missed an existing example in the docs.

**Comment by giswqs** [comment, marked as answer]
You can convert xarray to COG then use Map.add_geotiff() to display data. It doesn't support loading data directly from xarray in the memory yet.

https://leafmap.org/notebooks/47_numpy_to_cog/

**Reply by giswqs** [reply]
@banesullivan Does localtileserver support loading data in the memory, e.g., an xarray?

**Reply by banesullivan** [reply]
GRIB is supported by GDAL (not sure if there is a difference with GRIB2?).

I've been working with some GRIB data recently (again, not sure about any GRIB2 differences) and using GDAL/large-image to load them (FYI, localtileserver is a layer above large-image which in turn is a layer above GDAL). Here is an example:

Data from: https://www.ncei.noaa.gov/data/global-forecast-system/access/historical/analysis/

Specifically this dataset for the demo: https://www.ncei.noaa.gov/data/global-forecast-system/access/historical/analysis/200403/20040302/gfsanl_3_20040302_0000_000.grb

```py
from localtileserver import get_leaflet_tile_layer, TileClient
from ipyleaflet import Map

# Create a TileClient from a raster file
tile_client = TileClient('gfsanl_3_20040302_0000_000.grb')
```

```py
>>> tile_client.metadata()['bands'].keys()
dict_keys(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72', '73', '74', '75', '76', '77', '78', '79', '80', '81', '82', '83', '84', '85', '86', '87', '88', '89', '90', '91', '92', '93', '94', '95', '96', '97', '98', '99', '100', '101', '102', '103', '104', '105', '106', '107', '108', '109', '110', '111', '112', '113', '114', '115', '116', '117', '118', '119', '120', '121', '122', '123', '124', '125', '126', '127', '128', '129', '130', '131', '132', '133', '134', '135', '136', '137', '138', '139', '140', '141', '142', '143', '144', '145', '146', '147', '148', '149', '150', '151', '152', '153', '154', '155', '156', '157', '158', '159', '160', '161', '162', '163', '164', '165', '166', '167', '168', '169', '170', '171', '172', '173', '174', '175', '176', '177', '178', '179', '180', '181', '182', '183', '184', '185', '186', '187', '188', '189', '190', '191', '192', '193', '194', '195', '196', '197', '198', '199', '200', '201', '202', '203', '204', '205', '206', '207', '208', '209', '210', '211', '212', '213', '214', '215', '216', '217', '218', '219', '220', '221', '222', '223', '224', '225', '226', '227', '228', '229', '230', '231', '232', '233', '234', '235', '236', '237', '238', '239', '240', '241', '242', '243', '244', '245', '246', '247', '248', '249', '250', '251', '252', '253', '254', '255', '256', '257', '258', '259', '260', '261', '262', '263', '264', '265', '266', '267', '268', '269', '270', '271', '272', '273', '274', '275', '276', '277', '278', '279', '280', '281', '282', '283', '284'])
```


```py
# Create ipyleaflet TileLayer from that server
t = get_leaflet_tile_layer(tile_client, band=280, palette='viridis')

# Create ipyleaflet map, add tile layer, and display
m = Map(center=tile_client.center(), zoom=2)
m.add_layer(t)
m
```

![Screen Shot 2022-03-29 at 4 03 43 PM](https://user-images.githubusercontent.com/22067021/160714706-aa930194-6d91-4dab-85c4-848381f6e2a2.png)

**Reply by banesullivan** [reply]
Can confirm, GDAL supports GRIB2:

> GDAL supports GRIB1 (reading) and GRIB2 (reading and writing) format raster data ...

https://gdal.org/drivers/raster/grib.html

**Reply by banesullivan** [reply]
@giswqs, GRIB data should be handled out of the box with localtileserver and rasterio

**Reply by giswqs** [reply]
@banesullivan This is awesome!

@blaylockbk You can use `Map.add_local_tile()` to load GRIB data directly with only one line of code. 

```python
m = leafmap.Map()
data = 'gfsanl_3_20040302_0000_000.grb'
m.add_local_tile(data, band=280, palette='viridis')
m
```

![image](https://user-images.githubusercontent.com/5016453/160715652-351ea0c6-6bb1-4071-a4fd-fa0e2eec207e.png)

**Reply by blaylockbk** [reply]
Thanks for the information and quick response! I'm excited to give this a try.
